### PR TITLE
Extract c10d Store tests to dedicated test file

### PR DESCRIPTION
**Pull request by awgu**
Partially addresses #55340

**Overview**
This factors out `FileStoreTest`, `HashStoreTest`, `PrefixFileStoreTest`, `TCPStoreTest`, `PrefixTCPStoreTest`, `PythonStoreTest`, `RendezvousTest`, `RendezvousEnvTest`, `RendezvousFileTest`, and `RendezvousTCPTest` from `test_c10d_common.py` to a new file `test_store.py`.

Additionally, unused import/initialization statements are removed from `test_c10d_common.py`, and the minimal set of import/initialization statements are used for `test_store.py`. 

Also, this changes `.jenkins/pytorch/multigpu-test.sh`, `.jenkins/pytorch/win-test-helpers/test_distributed.bat`, and `test/run_test.py` to include the new `test_store.py`.

**Testing**
All commands shown are run on an AI AWS cluster. 

I check the Store tests:
```
python test/distributed/test_store.py
```

I also check `test_c10d_common.py` since it is the source of the refactored code. In addition, I check `test_c10d_nccl.py` and `test_c10d_gloo.py` since they import from `test_c10d_common.py`; those two should be the only test files depending on `test_c10d_common.py`.
```
python test/distributed/test_c10d_common.py 
python test/distributed/test_c10d_nccl.py
python test/distributed/test_c10d_gloo.py
```
`test_c10d_gloo.py` produces warnings about how using sparse tensors in TorchScript is experimental, but the warnings do not result from this PR's changes.

**Testing Issues** (To Be Revisited)
```
WORLD_SIZE=4 BACKEND=gloo gpurun pytest test/distributed/test_c10d_gloo.py
```
Running the above command fails three tests (written as `[Test]`: `[Error]`):
- `ProcessGroupGlooWrapperTest.test_collective_hang`: `RuntimeError: [../third_party/gloo/gloo/transport/tcp/pair.cc:598] Connection closed by peer [10.200.24.101]:15580`
- `CommTest.test_broadcast_coalesced_gloo_cuda`: `RuntimeError: cuda runtime error (3) : initialization error at ../aten/src/THC/THCGeneral.cpp:54`
- `CommTest.test_sequence_num_incremented_gloo_default`: `RuntimeError: cuda runtime error (3) : initialization error at ../aten/src/THC/THCGeneral.cpp:54`
However, running each of the following yields no errors:
```
WORLD_SIZE=4 BACKEND=gloo gpurun pytest test/distributed/test_c10d_gloo.py -k test_collective_hang
WORLD_SIZE=4 BACKEND=gloo gpurun pytest test/distributed/test_c10d_gloo.py -k test_broadcast_coalesced_gloo_cuda
WORLD_SIZE=4 BACKEND=gloo gpurun pytest test/distributed/test_c10d_gloo.py -k test_sequence_num_incremented_gloo_default
```
This suggests the existence of some inadvertent state dependency between tests (e.g. improper cleanup). I have not explored this further yet. In particular, I do not have a solid understanding of the tests to be able to explain why using `pytest` and `gpurun` induces the failure (since notably, running the `.py` directly shows no issue).

Similarly, running the following yields 47 errors:
```
WORLD_SIZE=4 BACKEND=nccl gpurun pytest test/distributed/test_c10d_nccl.py
```
The errors seem to all be simply complaining about the usage of `fork()` instead of `spawn()` for CUDA multiprocessing. Though, most of the tests in `test_c10d_nccl.py` ask for at least 2 CUDA devices, so I think that the `gpurun` is warranted (assuming that the test file does not need to be run partially on different machines). 

Both `test_c10d_common.py` and `test_store.py` work fine with `pytest`. 

**Other Notes**
I noticed that `torch.distributed` is imported both as `dist` and as `c10d` and that `c10d` is used throughout the Store tests. I was curious if this is intentional (as opposed to using `dist` to refer to `torch.distributed`). Also, the original [issue](https://github.com/pytorch/pytorch/issues/55340) suggests that the Store tests do not use multiprocessing, but I saw that `torch.multiprocessing` is still used in `TCPStoreTest`. 

The links for the Store files in the `CONTRIBUTING.md` [file](https://github.com/pytorch/pytorch/blob/master/torch/distributed/CONTRIBUTING.md) are broken. This can fixed in a separate PR.

### CODE REVIEW

File: torch/distributed/CONTRIBUTING.md [hunk 1]
```diff
@@ -73,12 +73,15 @@ A list of onboarding tasks can be found [here](https://github.com/pytorch/pytorc
...
+# Run the c10d unit tests.
 python test/distributed/test_c10d_common.py
 python test/distributed/test_c10d_gloo.py
 python test/distributed/test_c10d_nccl.py
 
-# Run distributed tests, including tests for Distributed Data Parallel
+# Run the Store tests.
+python test/distributed/test_store.py
+
+# Run distributed tests, including tests for Distributed Data Parallel.
```
Comment (andwgu) [inline torch/distributed/CONTRIBUTING.md:84]: Added a "." to be consistent.

**Comment by mrshenli** [comment]
> I noticed that torch.distributed is imported both as dist and as c10d and that c10d is used throughout the Store tests. I was curious if this is intentional (as opposed to using dist to refer to torch.distributed).

I don't think this is intentional. It might just because tests are added incrementally. And some PRs accidentally re-import `torch.distributed` using a different name.

**Comment by mrshenli** [comment]
> Also, the original issue suggests that the Store tests do not use multiprocessing, but I saw that torch.multiprocessing is still used in TCPStoreTest.

@rohan-varma might be referring to that store test does not use derive from `MultiProcessTestCase`. @rohan-varma  please correct me if I was wrong. 

https://github.com/pytorch/pytorch/blob/c03cae49fc70408a93f4de15a11a5a73997e9565/torch/testing/_internal/common_distributed.py#L371

File: test/distributed/test_store.py [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+import sys
+import tempfile
+import time
+import traceback
+import unittest
+from datetime import timedelta
+from sys import platform
+
+import torch
+import torch.distributed as c10d
```
Comment (mrshenli) [inline test/distributed/test_store.py:12]: nit: (as you mentioned, this is prior to this PR) Shall we add a followup PR to consolidate c10d and dist?

File: test/distributed/test_store.py [hunk 3]
```diff
@@ -0,0 +1,425 @@
+import os
```
Comment (mrshenli) [inline test/distributed/test_store.py:1]: Assuming this no major change in this file and it's mainly moving test contents around.

**Review by mrshenli — APPROVED** [review]
LGTM. Thanks for working on BE items. 

One thing I wanna mention is that, multigpu test only runs on master and won't run on PRs. It is possible to force CI to run all tests on PR with the following steps. 


1. create a branch on origin repo, its name must start with ci-all (say ci-all/andrew). This only needs to be done once. This is mine: https://github.com/pytorch/pytorch/tree/ci-all/mrshenli
2. force push the PR to that branch
3. create a PR from that ci-all/andrew branch to master. You can do this manually on github
4. that new PR should then run all ci-all tests

**Comment by mrshenli** [comment]
Could you please import this PR as a diff to trigger internal tests? You might need to add this new test to internal config files to enable it internally.

**Review by rohan-varma — APPROVED** [review]
Thanks for working on this!

> Running the above command fails three tests (written as [Test]: [Error]):

> ProcessGroupGlooWrapperTest.test_collective_hang: RuntimeError: [../third_party/gloo/gloo/transport/tcp/pair.cc:598] ?Connection closed by peer [10.200.24.101]:15580
> CommTest.test_broadcast_coalesced_gloo_cuda: RuntimeError: cuda runtime error (3) : initialization error at ../aten/src/THC/THCGeneral.cpp:54
> CommTest.test_sequence_num_incremented_gloo_default: RuntimeError: cuda runtime error (3) : initialization error at ../aten/src/THC/THCGeneral.cpp:54

I am planning to separate out the `ProcessGroupWrapper` tests into separate files as well, so hopefully that either resolves the issue or helps us get closer to understanding if one of these tests is truly broken/flaky.

**Pull request closed without merging**